### PR TITLE
wallet: Refactor and document CoinControl

### PR DESCRIPTION
**Pull request by aureleoules**
- Moves CoinControl function definitions from `coincontrol.h` to `coincontrol.cpp`
- Adds more documentation
- Renames class member for an improved comprehension
- Use `std::optional` for `GetExternalOutput`

### CODE REVIEW

**Comment by yancyribbens** [comment]
These changes improve the code quality and implement the recommended practices described in the [developer notes - source code organization](https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md?plain=1#L1051).

Concept ACK

**Review by w0xlt — APPROVED** [review]
ACK https://github.com/bitcoin/bitcoin/pull/26066/commits/ae67293025392e8fb5054b07a5173209c793ed6c

**Review by furszy — COMMENTED** [review]
In dba7f9b5:
Why only change `IsSelected` and `HasSelected` and not the setters? Should be consistent.

Saying that, even if we are consistent on the renaming, I'm not so sure about it. Do you think that `PreSelect(outpoint)`/`IsPreSelected(outpoint)` is more understandable than `Select(outpoint)`/`IsSelected(output)`?

At least in my view, it doesn't look more readable than before.
Would instead aim for a good docstring and done.

**Comment by aureleoules** [comment]
@furszy I've slightly improved the docstring of `Select` and `SelectExternal` and rolled back to `IsSelected` and `HasSelected`.
I kept the old `setSelected` as `m_preset_inputs` though.

Is it better?

**Review by furszy — COMMENTED** [review]
Yeah @aureleoules better 👍🏼 .

**Comment by aureleoules** [comment]
All cleaned-up @furszy!

File: src/wallet/coincontrol.h [hunk 1]
```diff
@@ -63,73 +63,56 @@ class CCoinControl
...
+     * Returns true if the given output is selected as an external input.
+     */
+    bool IsExternalSelected(const COutPoint& output) const;
+    /**
+     * Returns the external output for the given outpoint if it exists.
+     */
+    std::optional<CTxOut> GetExternalOutput(const COutPoint& outpoint) const;
+    /**
+     * Pre-selects the given output.
+     * The output will be included in the transaction even if it's not the most optimal choice.
```
Comment (furszy) [inline src/wallet/coincontrol.h:84]: nit: would remove the "pre-selects" term and go directly with something like "store the given outpoint."

I don't think that the "pre" term here provides information to readers. (the second paragraph is great to express what the process will do with the selected outpoint)

(same for the others methods as well).

**Review by furszy — APPROVED** [review]
ACK 9d4cfd6c

File: src/wallet/coincontrol.h (hunk 1 above)
Comment (aureleoules) [reply src/wallet/coincontrol.h:84]: How about this?
```suggestion
     * Lock-in the selected output for spending.
     * The output will be included in the transaction even if it's not the most optimal choice.
```

File: src/wallet/coincontrol.h [hunk 2]
```diff
@@ -63,76 +63,59 @@ class CCoinControl
...
+    bool HasInputWeight(const COutPoint& outpoint) const;
+    /**
+     * Returns the input weight.
+     */
+    int64_t GetInputWeight(const COutPoint& outpoint) const;
 
 private:
-    std::set<COutPoint> setSelected;
+    std::set<COutPoint> m_preset_inputs;
     std::map<COutPoint, CTxOut> m_external_txouts;
```
Comment (satsie) [inline src/wallet/coincontrol.h:119]: nit: you've done such a fabulous job documenting the header methods (no good deed goes unpunished!) that I think I understand an external input is one that does not belong to the wallet. 

If you find yourself in the code again, is there any chance to get a comment on this data structure describing what it is? Something like `//! Map of inputs that do not belong to the wallet` (if that is indeed the correct definition). Actually... just for my own learning, is that the correct definition?

File: src/wallet/coincontrol.h (hunk 1 above)
Comment (furszy) [reply src/wallet/coincontrol.h:84]: yeah, sounds good 👌🏼 .

**Comment by satsie** [comment]
ACK 9d4cfd6c5a40cafaeed31d0102b45a7e44dc0076, reviewed diff and ran unit + functional tests

I'm very new to the project but this was a great PR to review, especially after the changes that came in from earlier code review. Thanks for the comments you added on all the function definitions. Can confirm that they are helpful to new eyes and as far as I can tell, the changes are a nice cleanup that make things easier to read, understand, and follow the recommended practices (as mentioned in an earlier comment).

File: src/wallet/coincontrol.h (hunk 2 above)
Comment (aureleoules) [reply src/wallet/coincontrol.h:122]: Done!
And yes you are correct :) The wallet would not know how to sign these inputs but they are in the utxo set.

**Comment by aureleoules** [comment]
Thanks for the reviews, I addressed your comments @satsie @furszy.

**Review by furszy — APPROVED** [review]
ACK 287719a

**Comment by aureleoules** [comment]
sorry I failed my rebase @furszy

**Comment by furszy** [comment]
or well, not ACK :p.
You forgot to change the `m_preset_inputs` inside `CCoinControl`

**Comment by satsie** [comment]
ACK 5db911a369df3977c79e6d8f06ccc38551db9efb

**Review by furszy — APPROVED** [review]
ACK 5db911a3

**Comment by aureleoules** [comment]
I think this is ready for merging? @w0xlt if you could reACK that would be great :smile:.

**Comment by aureleoules** [comment]
Rebased.

**Review by theStack — APPROVED** [review]
Code-review ACK 22801dff82d5d5329a75cb7fc14bd6e05dc67a5b

Nice cleanup and documentation improvement!
nit: I have one more suggestion for an improvement, the interface of `CCoinControl::ListSelected` would be more straight-forward if the result is returned directly instead of using an out-parameter, simplifying the caller-side:

<details>
  <summary>Diff</summary>

```diff
diff --git a/src/qt/coincontroldialog.cpp b/src/qt/coincontroldialog.cpp
index bd9a90a89..4b9ed9028 100644
--- a/src/qt/coincontroldialog.cpp
+++ b/src/qt/coincontroldialog.cpp
@@ -412,8 +412,7 @@ void CoinControlDialog::updateLabels(CCoinControl& m_coin_control, WalletModel *
     unsigned int nQuantity      = 0;
     bool fWitness               = false;

-    std::vector<COutPoint> vCoinControl;
-    m_coin_control.ListSelected(vCoinControl);
+    auto vCoinControl{m_coin_control.ListSelected()};

     size_t i = 0;
     for (const auto& out : model->wallet().getCoins(vCoinControl)) {
diff --git a/src/wallet/coincontrol.cpp b/src/wallet/coincontrol.cpp
index 9837bb442..009e95c2c 100644
--- a/src/wallet/coincontrol.cpp
+++ b/src/wallet/coincontrol.cpp
@@ -58,9 +58,9 @@ void CCoinControl::UnSelectAll()
     m_selected_inputs.clear();
 }

-void CCoinControl::ListSelected(std::vector<COutPoint>& vOutpoints) const
+std::vector<COutPoint> CCoinControl::ListSelected() const
 {
-    vOutpoints.assign(m_selected_inputs.begin(), m_selected_inputs.end());
+    return {m_selected_inputs.begin(), m_selected_inputs.end()};
 }

 void CCoinControl::SetInputWeight(const COutPoint& outpoint, int64_t weight)
diff --git a/src/wallet/coincontrol.h b/src/wallet/coincontrol.h
index 0ff2ced72..953e20ab4 100644
--- a/src/wallet/coincontrol.h
+++ b/src/wallet/coincontrol.h
@@ -100,7 +100,7 @@ public:
     /**
      * List the selected inputs.
      */
-    void ListSelected(std::vector<COutPoint>& vOutpoints) const;
+    std::vector<COutPoint> ListSelected() const;
     /**
      * Set an input's weight.
      */
diff --git a/src/wallet/spend.cpp b/src/wallet/spend.cpp
index 2da95f037..c5deb8d33 100644
--- a/src/wallet/spend.cpp
+++ b/src/wallet/spend.cpp
@@ -145,9 +145,7 @@ util::Result<PreSelectedInputs> FetchSelectedInputs(const CWallet& wallet, const
                                             const CoinSelectionParams& coin_selection_params) EXCLUSIVE_LOCKS_REQUIRED(wallet.cs_wallet)
 {
     PreSelectedInputs result;
-    std::vector<COutPoint> vPresetInputs;
-    coin_control.ListSelected(vPresetInputs);
-    for (const COutPoint& outpoint : vPresetInputs) {
+    for (const COutPoint& outpoint : coin_control.ListSelected()) {
         int input_bytes = -1;
         CTxOut txout;
         if (auto ptr_wtx = wallet.GetWalletTx(outpoint.hash)) {
```
</details>

**Comment by aureleoules** [comment]
Thanks for the suggestion @theStack. I pushed b76fbbe8bf9504d8c08cc2f1281782db5393c02d and gave you authorship.

**Review by theStack — APPROVED** [review]
ACK b76fbbe8bf9504d8c08cc2f1281782db5393c02d

**Comment by furszy** [comment]
While I would like to merge this one (it's good as is), I think that would be better to check #25273 first. Mainly because this PR is about improving the `CoinControl` class documentation and #25273 modifies a big part of it (which, if we merge this one first, will make us re-do this doc improvements in another follow-up later).

**Comment by furszy** [comment]
@MarcoFalke, guess that Drahtbot shouldn't have requested review to me.

**Comment by MarcoFalke** [comment]
Yeah, if it was live 3 weeks ago, it would have requested a review from you before your comment. Which, I guess, would have made sense.

Also, it will only request reviews when there is an email notification sent out anyway, so the harm/risk should be minimal. But let me know if it should be tamed down or turned off.

**Comment by achow101** [comment]
Are you still working on this?

**Comment by aureleoules** [comment]
Rebased

**Review by murchandamus — APPROVED** [review]
ACK daba95700b0b77a2e898299f218c47a69ed2c7d0

**Comment by achow101** [comment]
ACK daba95700b0b77a2e898299f218c47a69ed2c7d0

**Pull request merged**